contracts/Vault.vy

```diff
@@ -1382,3 +1397,53 @@ def sweep(_token: address, _value: uint256 = MAX_UINT256):
     if _value == MAX_UINT256:
         value = ERC20(_token).balanceOf(self)
     self.erc20_safe_transfer(_token, self.governance, value)
+
+
+@view
+@internal
+def permit_digest(owner: address, spender: address, amount: uint256, nonce: uint256, expiry: uint256) -> bytes32:
+    return keccak256(
+        concat(
+            b'\x19\x01',
+            self.DOMAIN_SEPARATOR,
+            keccak256(
+                concat(
+                    PERMIT_TYPE_HASH,
+                    convert(owner, bytes32),
+                    convert(spender, bytes32),
+                    convert(amount, bytes32),
+                    convert(nonce, bytes32),
+                    convert(expiry, bytes32),
+                )
+            )
+        )
+    )
+
+
+@external
+def permit(owner: address, spender: address, amount: uint256, nonce: uint256, expiry: uint256, signature: Bytes[65]) -> bool:
+    """
+    @notice
+        Approves spender by owner's signature to expend owner's tokens.
+        See https://eips.ethereum.org/EIPS/eip-2612.
+    
+    @param owner The address which is a source of funds and has signed the Permit.
+    @param spender The address which is allowed to spend the funds.
+    @param amount The amount of tokens to be spent.
+    @param nonce The current nonce of the owner.
+    @param expiry The timestamp after which the Permit is no longer valid.
+    @param signature A valid secp256k1 signature of Permit by owner encoded as r, s, v.
+    """
+    assert owner != ZERO_ADDRESS  # dev: invalid owner
+    assert expiry >= block.timestamp  # dev: permit expired
+    assert nonce == self.nonces[owner]  # dev: invalid nonce
+    digest: bytes32 = self.permit_digest(owner, spender, amount, nonce, expiry)
+    # NOTE: signature is packed as r, s, v
+    r: uint256 = convert(slice(signature, 0, 32), uint256)
```

Comment:
is it better to just expect these as args like :
https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2ERC20.sol#L81
?

Reply:
Vyper can't do `uint8`, so we can't match their signature anyway. See how it's passed in the tests to see why this approach is viable, you can pass the signature arguments from `eth_accounts` signed message as is. I haven't checked if it's the case with web3.js and ethers.js.

Reply:
yeah web3 and ethers have split signature, but if vyper doesnt have uint8 yeah its better to do it this way

Reply:
I actually want to update `ecrecover` to except either `Bytes[65]` vrs blob, or separate v/r/s integers. 3 integers is more efficient though.